CMakeLists.txt

```diff
@@ -34,18 +34,24 @@ else()
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wextra -Wpedantic -Wundef -Wshadow")
     # TODO: Remove this when fixed
     if(ANDROID)
-        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-maybe-uninitialized -Wno-uninitialized")
+        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-uninitialized")
     elseif(${CMAKE_CXX_COMPILER_ID} MATCHES ".*[Cc]lang")
         set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wunreachable-code -Wshorten-64-to-32 -Wold-style-cast -Wconditional-uninitialized -Wextra-semi -Wno-nested-anon-types")
     endif()
 endif()
 
+if(ANDROID)
+    # On some architectures char is unsigned by default. Make it signed
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fsigned-char")
```

Comment:
It seems that quite a lot of flags were hiding as defaults. Comparing compiler flags between `r10e` and `r18b` revealed the following differences:

**NDK18**
```
LANGUAGE_COMPILE_FLAGS=
-D__ANDROID_API__=21
-DNDEBUG
-fexceptions
-frtti
-fsigned-char
-funwind-tables
-no-canonical-prefixes
-O3
-Wall
-Wextra
-Wno-uninitialized
-Wpedantic
-Wshadow
-Wundef
```

**NDK10:**
```
-DNDEBUG
-fdata-sections
-fexceptions
-ffunction-sections
-finline-limit=300
-fomit-frame-pointer
-fPIC
-frtti
-fsigned-char
-fstrict-aliasing
-funswitch-loops
-funwind-tables
-fvisibility=hidden
-no-canonical-prefixes
-Os
-std=c++14
-Wa,--noexecstack
-Wall
-Wextra
-Wno-maybe-uninitialized
-Wno-psabi
-Wno-uninitialized
-Wpedantic
-Wshadow
-Wundef
```

Especially lacking `-ffunction-sections` and `-fdata-sections` could go a long to explaining way the final Java .so file suddenly got a lot bigger.

Reply:
Which of these flags would you like me to apply?

Reply:
@fealebenpae I'll try to iterate locally and then report back as some of them probably also impact performance, which we need to check as well. 

But we also need to verify exactly which flags OpenSSL is built with. Do you know if that is easy to check?

Reply:
These are the flags the OpenSSL Configure script builds with for android:
```
--sysroot=/opt/android-ndk-r18b/platforms/android-21/arch-arm64
-c
-DNDEBUG
-DOPENSSL_PIC
-DOPENSSL_USE_NODELETE
-D__ANDROID_API__=21
-fPIC
-MF crypto/bio/b_print.d.tmp
-MMD
-MT crypto/bio/b_print.o
-O3
-pthread
-target aarch64-linux-android
-Wall
```
, or the complete clang invocation:
```
clang  -I. -Icrypto/include -Iinclude -fPIC -pthread  -target aarch64-linux-android -gcc-toolchain /opt/android-ndk-r18b/toolchains/aarch64-linux-android-4.9/prebuilt/darwin-x86_64 --sysroot=/opt/android-ndk-r18b/platforms/android-21/arch-arm64 -Wall -O3 -fPIC -DOPENSSL_USE_NODELETE -DOPENSSL_PIC -DOPENSSLDIR="\"/Users/yavorgeorgiev/android/arm64/ssl\"" -DENGINESDIR="\"/Users/yavorgeorgiev/android/arm64/lib/engines-1.1\"" -D__ANDROID_API__=21 -isystem /opt/android-ndk-r18b/sysroot/usr/include/aarch64-linux-android -isystem /opt/android-ndk-r18b/sysroot/usr/include -DNDEBUG -D__ANDROID_API__=21 -MMD -MF crypto/bio/b_print.d.tmp -MT crypto/bio/b_print.o -c -o crypto/bio/b_print.o crypto/bio/b_print.c
```

I could probably add or change flags here and rebuild, just let me know what the flags should be.

Reply:
Most important is to add `-ffunction-sections` and `-fdata-sections`, but it would also be interesting to see what happens if we replace `-O3` with `-Os`.

Currently we are also building Core/Java with `-fsigned-char`, but I don't know if that is required for OpenSSL. Since the unit test parses, I guess not, but then again, I don't claim to fully understand the implications of most of these build flags.